### PR TITLE
Issues 8714: HttpClientIntroductionAdvice Subscriber may not have complete called by some Publisher implementations

### PR DESCRIPTION
**Pull request by rawilder**
I manually cancel the subscription as is dictated by the reactive streams spec, a subscriber must cancel when it does not need items anymore, and then completes on its own. Added a test for the specific bug I was seeing in the kotlin test suite.

I feel like this will need some fine tuning for edge cases like onComplete being called twice, perhaps by inlining the onComplete functionality and making onComplete a NOP

Unfortunately I was not able to build locally, I kept getting
```
An exception occurred applying plugin request [id: 'io.micronaut.build.internal.dependency-updates']
> Failed to apply plugin class 'io.micronaut.build.catalogs.MicronautVersionCatalogUpdatePlugin'.
   > Cannot add task 'dependencyUpdates' as a task with that name already exists.
```

Motivation for why here instead of kotlinx.coroutines:
https://github.com/Kotlin/kotlinx.coroutines/issues/3608#issuecomment-1416156578
https://github.com/reactive-streams/reactive-streams-jvm/issues/543

### CODE REVIEW

**Comment by rawilder** [comment]
I read through the errors and am fairly certain they are not caused by my change so I am marking this ready for review.

**Comment by yawkat** [comment]
The test does not work. It passes even without the change you made to HttpClientIntroductionAdvice. Please take a look at it.

**Comment by rawilder** [comment]
Hmm interesting. I am trying to replicate the behavior and I am also having trouble. My original issue example was 3.8.x so I will try rebasing and see if I replicate it. So far it looks like some intermediary subscriber is requesting 32 instead of the 1 I was seeing with my example project. Will test and see what I find.

**Comment by rawilder** [comment]
That's exactly what it was. I am going to rebase this PR against 3.8.x. I had just assumed that 3.7.x was the culprit as it had the large client rewrite. My mistake!

Failed test:
image

Modification to cause failure:
image

**Review by yawkat — APPROVED** [review]

**Comment by yawkat** [comment]
ive changed it slightly to match the MonoNext implementation, and removed the unnecessary atomicref

File: test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientSpec.kt [hunk 1]
```diff
@@ -42,4 +43,14 @@ class SuspendClientSpec {
...
     }
 
+    @Test
+    fun testFlowAsPublisherInFilterStep() {
+        val server = ApplicationContext.run(EmbeddedServer::class.java, mapOf("spec.name" to "SuspendClientSpec"))
+        val ctx = server.applicationContext
+        val response = runBlocking {
+            ctx.getBean(SuspendClient::class.java).call(SuspendClientFilter.filterCheck)
+        }
+        Assertions.assertEquals(response, "{\"newState\":\"${SuspendClientFilter.filterCheck}\"}")
```
Comment (sdelamo) [inline test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientSpec.kt:55]: we should close the server at the end of the test with `server.close()`

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import kotlinx.coroutines.withContext
+import org.reactivestreams.Publisher
+
+@Filter
+class SuspendClientFilter : HttpClientFilter {
+    override fun doFilter(request: MutableHttpRequest<*>, chain: ClientFilterChain): Publisher<out HttpResponse<*>> {
+        // if request contains filterCheck, then do flow step, else proceed
+        return if ((request.body.orElse(null) as? Map<*, *>)?.containsValue(filterCheck) == true) {
+            flow { emit(getValue()) }.flatMapMerge {
+                chain.proceed(request.header("myTestHeader", it)).asFlow()
```
Comment (sdelamo) [inline test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]: Are we doing anything with the HTTP Header `myTestHeader`. Should we test it?

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientSpec.kt (hunk 1 above)
Comment (sdelamo) [reply test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientSpec.kt:55]: I added a [commit to close the server](https://github.com/micronaut-projects/micronaut-core/pull/8715/commits/09b140d38ef3ff5fae5a3bfb130d038a6c8f4395)

File: test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt (hunk 2 above)
Comment (rawilder) [reply test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]: It is simply a usage of the value from the flow publisher. It is not necessary to test. We could easily emit and using nothing and it would have the same effect I imagine.

Comment (yawkat) [reply test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]: ok i think it's fine as-is

Comment (sdelamo) [reply test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]: if we don't verify anything, I would prefer to replace: 

```java
 chain.proceed(request.header("myTestHeader", it)).asFlow()
```

with: 

```java
 chain.proceed(request).asFlow()
```

thoughts?

Comment (sdelamo) [inline test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]:
```suggestion
                chain.proceed(request).asFlow()
```

**Review by sdelamo — APPROVED** [review]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt (hunk 2 above)
Comment (rawilder) [reply test-suite-kotlin/src/test/kotlin/io/micronaut/http/client/SuspendClientFilter.kt]: I see it's already merged but as long as FlowAsPublisher is somewhere in the chain it should cause the behavior.
